contrib/epee/include/net/http_auth.h

```diff
@@ -45,15 +45,22 @@ namespace net_utils
     public:
       struct login
       {
-        login() = delete;
+        login() : username(), password() {}
+        login(std::string username_, std::string password_)
+          : username(std::move(username_)), password(std::move(password_))
```

Comment:
I think pass-by-value is preferable in this situation. This [blog post by Dave Abrahams](http://web.archive.org/web/20140205194657/http://cpp-next.com/archive/2009/08/want-speed-pass-by-value/) changed my view on pass-by-value years back. Compilers are really "smart". The one that has changed since that post is C++11 - if RVO is not possible, a C++11 compiler is required to do a move if possible.

An alternative would be to have `const std::string&` and `std::string&&` constructors, then have the former call the latter via delegated constructor. But with two arguments, 4 constructors are needed for all permutations. This implementation covers all of those cases, and is identical to the aggregate initialization I was doing before (with slightly more boilerplate).

There are two downsides: (1) the type cannot be forward-declared, and (2) every caller that needs to copy/move has to "call" the appropriate constructor first, instead of the callee doing it once (small code bloat for function call). Dave Abrahams discusses both in that post. (1) Isn't an issue with `std::string` since you cannot realistically forward-declare it. And for (2), in some cases the compiler applies RVO, and neither a move or copy is done anyway.